stdlib/public/runtime/Exclusivity.cpp

```diff
@@ -0,0 +1,240 @@
+#define SWIFT_EXCLUSIVITY_USE_PTHREAD_SPECIFIC 0
+#endif
+
+#endif
+
+#if SWIFT_EXCLUSIVITY_USE_PTHREAD_SPECIFIC
+#include <pthread.h>
+#include <stdio.h>
+#endif
+
+using namespace swift;
+
+bool swift::_swift_disableExclusivityChecking = false;
+
+static const char *getAccessName(ExclusivityFlags flags) {
+  switch (flags) {
+  case ExclusivityFlags::Read: return "read";
+  case ExclusivityFlags::Modify: return "modify";
+  }
+  return "unknown";
+}
+
+namespace {
+
+/// A single access that we're tracking.
+struct Access {
+  void *Pointer;
+  uintptr_t HereAndAction;
+  Access *Next;
+
+  enum : uintptr_t {
+    ActionMask = (uintptr_t)ExclusivityFlags::ActionMask,
+    HereMask = ~ActionMask
+  };
+
+  Access **getHere() const {
+    return reinterpret_cast<Access**>(HereAndAction & HereMask);
+  }
+  ExclusivityFlags getAccessAction() const {
+    return ExclusivityFlags(HereAndAction & ActionMask);
+  }
+
+  void initialize(void *pointer, Access **here, ExclusivityFlags action) {
+    assert(*here == nullptr && "not inserting to end of list");
+    Pointer = pointer;
+    HereAndAction = reinterpret_cast<uintptr_t>(here) | uintptr_t(action);
+    Next = nullptr;
+  }
+};
+
+static_assert(sizeof(Access) <= sizeof(ValueBuffer) &&
+              alignof(Access) <= alignof(ValueBuffer),
+              "Access doesn't fit in a value buffer!");
+
+/// A set of accesses that we're tracking.  Just a singly-linked list.
+class AccessSet {
+  Access *Head = nullptr;
+public:
+  constexpr AccessSet() {}
+
+  void insert(Access *access, void *pointer, ExclusivityFlags flags) {
+    auto action = getAccessAction(flags);
+
+    Access **curP = &Head;
+    for (Access *cur = *curP; cur != nullptr; curP = &cur->Next, cur = *curP) {
+      // Ignore accesses to different values.
+      if (cur->Pointer != pointer)
+        continue;
+
+      // If both accesses are reads, it's not a conflict.
+      if (action == ExclusivityFlags::Read &&
+          action == cur->getAccessAction())
+        continue;
+
+      // Otherwise, it's a conflict.
+      // TODO: try to recover source-location information from the return
+      // address.
+      fprintf(stderr,
+              "%s/%s access conflict detected on address %p, aborting\n",
+              getAccessName(action), getAccessName(cur->getAccessAction()),
+              pointer);
+      abort();
+    }
+
+    access->initialize(pointer, curP, action);
+  }
+
+  static void remove(Access *access) {
+    *access->getHere() = access->Next;
+  }
+};
+
+/// A set of independent access sets.  This is not designed to put
+/// the access sets on different cache lines, so it's fine for
+/// thread-local sets and probably not fine for concurrent sets.
+class AccessSets {
+  enum { NumAccessSets = 8 };
+  AccessSet Sets[NumAccessSets];
+
+public:
+  constexpr AccessSets() = default;
+  AccessSets(const AccessSets &) = delete;
+  AccessSets &operator=(const AccessSets &) = delete;
+
+  AccessSet &get(void *pointer) {
```

Comment:
This hash function looks expensive, but that can be left to future performance optimization.